smartconfig/entry.py

```diff
@@ -0,0 +1,138 @@
+from typing import Any, Dict, NoReturn, Optional, Tuple
+
+from smartconfig import _registry
+from smartconfig._registry import configuration_for_module, global_configuration, lookup_global_configuration
+from smartconfig.exceptions import ConfigurationKeyError, InvalidOperation, PathConflict
+
+
+class _ConfigEntryMeta(type):
+    """
+    Metaclass used to define special ConfigEntry behaviors.
+
+    Note: Using this metaclass outside of the library is currently not supported.
+
+    The lookup of attributes can be customized by subclassing this metaclass and changing `_get_attribute_path`.
+    """
+
+    def __getattribute__(cls, name: str) -> Optional[Any]:
+        """
+        Look up the attribute through the configuration system.
+
+        If the attribute name starts with `_`, normal lookup is done, otherwise _registry.global_configuration is used.
+
+        Args:
+            name: Attribute to lookup
+
+        Raises:
+            ConfigurationKeyError: The attribute doesn't exist.
+        """
+        # Use the normal lookup for attribute starting with `_`
+        if name.startswith('_'):
+            return super().__getattribute__(name)
+
+        path, attribute = cls._get_attribute_path(name)
+
+        return lookup_global_configuration(path, attribute)
+
+    def __new__(cls, name: str, bases: Tuple[type, ...], dict_: Dict[str, Any], path: Optional[str] = None) -> type:
+        """
+        Add the `__defined_attributes` and `__path_override` attributes to the new entry.
+
+        Args:
+            path: Custom path used for this entry instead of the module name.
+        """
+        names = chain(dict_.keys(), dict_.get('__annotations__', {}).keys())
+        dict_[f"{cls.__name__}__defined_attributes"] = {name for name in names if not name.startswith('_')}
+
+        dict_[f"{cls.__name__}__path_override"] = path
+
+        return super().__new__(cls, name, bases, dict_)
+
+    def _register_entry(cls) -> None:
+        """Set the `__path` attribute and register the entry."""
+        cls.__path = cls.__path_override or cls.__module__
+        if cls.__path in _registry.configuration_for_module:
+            raise PathConflict(f"An entry at {cls.__path!r} already exists.")  # TODO: Add an FAQ link.
+
+        configuration = {
+            key: value for key, value in cls.__dict__.items() if not key.startswith('_')
+        }
+
+        if cls.__path not in global_configuration:
+            global_configuration[cls.__path] = configuration
+        # We already have some overrides for this path.
+        else:
+            for key, value in configuration.items():
+                path, key = cls._get_attribute_path(key)
+
+                if path not in global_configuration:
+                    global_configuration[path] = {}
+
+                # We only write values that aren't already defined.
+                if key not in global_configuration[path]:
+                    global_configuration[path][key] = value
```

Comment:
So we would have a function running for duplicates attributes? I’m not sure if it is a good idea, since we could have some issues with user defined code overriding values coming from configuration files, that would be a pretty weird behavior.

Reply:
Solved by removing `_get_attribute_path`.